### PR TITLE
Add a JIT interpreter

### PR DESCRIPTION
**Pull request by zdevito**
The separate interpreter is used to graphs with a lower overhead than
converting them to autograd graphs. Some notes:

* does not support Handles/PythonOp/CppOp, these will be in a future commit
* jit_closure.cpp still exists and we fall back to it for now when
  cannot handle something because of PythonOp/CppOp
* In order to support retain_graph=True, the interpreter can be cloned,
  creating a copy that can be run with different arguments. This is
  assumed to be the non-standard case so cloning is not particularly optimized.
  No tensor _data_ is copied, but the at::Tensor list in the interpreter is.
  If we hit problems, there is a lot we could do (such as register allocation)
  to minimize the stuff that needs to be copied.
* Uses a pImpl pattern to keep implementation details out of its header file.
* Modifies the way getTensorOp works so that it reads/writes to already-existing
  vectors, this prevents needing to realloc these buffers each time.
* Timings are here: https://gist.github.com/zdevito/5a20ac29fb1b9e449e693b67dc478127
  This reduces overhead to about the same as running it in python.
  It is about 10us faster to run the same thing using ATen directly.

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -204,7 +204,9 @@ static variable_list call_function(FunctionTask& task) {
     auto& callback = it_p.first->second;
     if (!callback(&fn, inputs)) return variable_list(fn.next_functions.size());
   }
-
+  if(!task.base->keep_graph) {
```
Comment (zdevito) [inline torch/csrc/autograd/engine.cpp]: The interpreter needs to know _before_ it executes that it should be releasing variables. If you don't like this way, I am open to others, but it does need to know ahead of time for efficient execution and small memory footprint.

File: torch/csrc/jit/interpreter_autograd_function.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+struct InterpreterAutogradFunction : public autograd::Function {
+  InterpreterAutogradFunction(const jit::Function & function)
+  : interp_(function) {}
+  InterpreterAutogradFunction(const Interpreter & interp_, autograd::FunctionFlags && f)
+  : autograd::Function(std::move(f)), interp_(interp_) {}
+
+  virtual void willReleaseVariables() override {
+    keep_graph = false;
+  }
+  virtual autograd::variable_list apply(const autograd::variable_list& inputs) override {
```
Comment (zdevito) [inline torch/csrc/jit/interpreter_autograd_function.h]: I a not confident that Variable flags are being propagated correctly as the interpreter runs. @apaszke can you let me know what I am suppose to be doing with them here. I imagine we should be just saving the flags from the trace and reproducing them here.

File: torch/csrc/jit/interpreter.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    return pImpl != nullptr;
+  }
+private:
+  std::shared_ptr<FunctionImpl> pImpl;
+  friend class InterpreterImpl;
+};
+
+struct Interpreter {
+  Interpreter(const Function & function);
+  void runOneStage(
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.h]: You can tell this is the case when you look at the 'clone' method below, but there should be comments here saying that an interpreter is a mutable resource which is not thread safe and can only be run once. Perhaps there is another name rather than Interpreter which expresses this more clearly, like "InterpreterHandle"

File: torch/csrc/jit/interpreter.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+// The interpreter run Graphs with Tensor inputs and Tensor outputs
+// a separate component in the autograd handles unwrapping and wrapping
+// variable objects for use in the interpreter.
+
+struct FunctionImpl;
+struct InterpreterImpl;
+struct Graph;
+
+struct Function {
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.h]: Blah, this is the same name as autograd Function, which will make it harder to grep for instances of this particular concept. How about `GraphBCO` (graph bytecode object).

File: torch/csrc/jit/interpreter.cpp [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    return [fusion_fn](InputList inputs, OutputList outputs) {
+      fusion_fn->launch(inputs, outputs);
+    };
+#else
+    throw std::runtime_error("don't know how to execute FusionGroups without CUDA");
+#endif
+  IR_ELSEIF(Constant)
+    auto t = value->t(kvalue);
+    return [t](InputList inputs, OutputList outputs) {
+      outputs.push_back(t);
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.cpp]: Err... do we allow inplace operations in the new interpreter? If so, wouldn't we get in trouble if we did an inplace op on this constant? I remember thinking we had this hazard in the old code too, but we had a clone to prevent this from happening:

```
struct WrapConstant : public Function {
  WrapConstant(at::Tensor value)
    : value(std::move(value)) {
    is_executable = true;
    num_inputs = 1;
  }

  virtual variable_list apply(const variable_list& inputs) {
    if (inputs.size() != 1 || inputs[0].defined())
      throw std::logic_error("WrapConstant nodes should only receive a single NULL input");
    AutoGPU guard(value);
    return {make_variable(value.clone())};
  }

  at::Tensor value;
};
```

File: torch/csrc/jit/interpreter.cpp [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    return getTensorOp(node).op;
+  IR_END()
+}
+
+
+// We need some lists for inputs and outputs. To keep all the memory
+// contiguous we allocate a single vector and use offsets into the vector
+// which are stored in the RegList struct
+// start is an offset into int_data of Function if this list is integers
+// and bool_data of Function if this list is booleans (only free_flags)
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.cpp]: I'm annoyed that there is no static or dynamic information telling me whether or not a RegList struct indexes into int_data or bool_data. Can we at least get comments saying which array we're indexing into?

Comment (ezyang) [reply torch/csrc/jit/interpreter.cpp]: (The overengineer in me wonders if you couldn't use a custom allocator to get everything contiguous... probably not worth it haha)

File: torch/csrc/jit/interpreter.cpp [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+struct RegList {
+  int start;
+  int size;
+};
+
+struct UseList {
+  // values to be used
+  RegList values;
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.cpp]: What is a "value"? (doc request)

File: torch/csrc/jit/interpreter.cpp [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+      stages.emplace_back();
+      auto & stage = stages.back();
+      intListBegin(stage.inputs);
+      for(;input_pos < graph->inputs().size(); input_pos++) {
+        auto input = graph->inputs()[input_pos];
+        if((int64_t)input->stage() > cur_stage)
+          break;
+        // unused inputs are given a false register -1 so that we never hold a
+        // reference to the tensor data, otherwise we would fail to clean them
+        // up since they do not have a last use at which to free them
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.cpp]: Nice.

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/autograd/engine.cpp]: This seems fine to me.

File: torch/csrc/jit/interpreter.cpp (hunk 6 above)
Comment (ezyang) [reply torch/csrc/jit/interpreter.cpp]: After reading more code, I also wonder if this isn't misnamed. RegList struct is being used to universally represent "array in contiguous sequence of memory", it really isn't a "list of registers" at all.

File: torch/csrc/jit/interpreter.cpp [hunk 9]
```diff
@@ -0,0 +1,288 @@
+#include "interpreter.h"
+#include "torch/csrc/jit/ir.h"
+#include "torch/csrc/jit/generated/aten_dispatch.h"
+#ifdef WITH_CUDA
+#include "torch/csrc/jit/fusion_compiler.h"
+#endif
+
+namespace torch { namespace jit {
```
Comment (ezyang) [inline torch/csrc/jit/interpreter.cpp]: It'd be positively lovely if you could record down which perf optimizations you *know* pushed the needle down, if you remember this from when you were working on the interpreter. (Unless, well, it sprang fully formed from your head as fast as it is today. That's cool too 8)

File: tools/jit/gen_jit_dispatch.py [hunk 10]
```diff
@@ -85,16 +85,16 @@ def is_tensor_arg(arg):
         if 'namespace' in decl['method_of']:
             if any(arg['simple_type'] == 'TensorList' for arg in arguments):
                 assert sum(map(is_tensor_arg, arguments)) == 1
-                args = ['as_tensor_list(vars)' if is_tensor_arg(arg) else arg['name']
+                args = ['inputs' if is_tensor_arg(arg) else arg['name']
```
Comment (ezyang) [inline tools/jit/gen_jit_dispatch.py]: No one expects the Spanish Implicit Conversion!!

File: torch/csrc/jit/interpreter_autograd_function.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/jit/interpreter_autograd_function.h]: Actually, assuming that you've accurately got all the inputs, I think this actually will DTRT.

**Review by ezyang — APPROVED** [review]
This is so groovy! As usual, you should make the final call whether or not you want to address the comments I made or not; the only outright potential bug is us not cloning tensor constants from the graph IR when we construct an interpreter state.

File: torch/csrc/jit/interpreter.h (hunk 4 above)
Comment (ezyang) [reply torch/csrc/jit/interpreter.h]: (Yes, I know we don't actually have a bytecode today. But we might, someday...)

File: torch/csrc/jit/interpreter.cpp [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+      auto & inst = stages.back().instructions.back();
+      intListBegin(inst.inputs.values);
+      for(auto input : node->inputs()) {
+        intListInsert(inst.inputs.values, getOrAllocateRegister(input, true));
+      }
+      intListBegin(inst.outputs);
+      for(auto output : node->outputs()) {
+        intListInsert(inst.outputs, getOrAllocateRegister(output));
+      }
+      inst.callback = getCallback(node);
```
Comment (apaszke) [inline torch/csrc/jit/interpreter.cpp]: Wouldn't it be more natural to just do
```cpp
Instruction inst;
// initialize everything here
stages.back().instructions.push_back(std::move(inst));
```

File: torch/csrc/jit/interpreter.cpp [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    }
+    insertStagesTo(cur_stage, graph->stage(), input_pos, output_pos);
```
Comment (apaszke) [inline torch/csrc/jit/interpreter.cpp]: Why is this necessary?

**Review by apaszke — APPROVED** [review]
This looks really good. I left some questions. In general I think the code could be simplified a bit (especially `RegList`s are quite unnatural, but I can address that in a next PR).

My biggest concern is the pImpl pattern. Why do we need this? We only have a single implementation, so why add the level of indirection?

File: torch/csrc/jit/interpreter_autograd_function.h (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/interpreter_autograd_function.h]: No, this won't work. We have the [information necessary to reproduce the flags in `TracingState`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tracer_state.h#L68), so it's possible to do right now. I'm fine with merging as is and I can fix that later.

File: torch/csrc/jit/interpreter_autograd_function.h [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    std::vector<at::Tensor> tinputs;
+    std::vector<at::Tensor> toutputs;
+    for(auto & i : inputs) {
+      tinputs.push_back(i.data());
+    }
+    Interpreter interp = (keep_graph) ? interp_.clone() : interp_;
```
Comment (apaszke) [inline torch/csrc/jit/interpreter_autograd_function.h]: This will call the copy constructor of `interp` - probably not what you wanted.

File: torch/csrc/jit/interpreter.cpp (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: I am assuming the Graph doesn't have any inplace operators. What we can do later is optimize the instruction generation to re-use inputs tensors when an inplace variant exists and it is safe to do (no more uses, not ruining a constant). Copying constants can potentially add huge overhead, so it makes sense to stick with working with the fast path.

File: torch/csrc/jit/interpreter.cpp (hunk 9 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: I didn't add optimizations after the fact. It's easier to start with good perf than try to optimize later. Instead there were two goals:

1. Don't allocate memory when you running (there is a single malloc when the Interpreter is created)
2. Scan linearly through memory as much as possible to be friendly to cache.

It is not clear if either of these were necessary but in some sense it is easier writing it this way, then writing it another way and then trying all optimizations.

File: tools/jit/gen_jit_dispatch.py (hunk 10 above)
Comment (zdevito) [reply tools/jit/gen_jit_dispatch.py]: I changed all of these methods to actually work Tensor -> Tensor rather than Variable -> Variable. So this is no longer a cast.

File: torch/csrc/jit/interpreter.cpp (hunk 6 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: Yeah, we can probably add some more infra to make this clearer.

File: torch/csrc/jit/interpreter.h (hunk 3 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter.h]: I'll add some comments. I will probably rename Interpreter -> InterpreterState to make it more clear. `lua_State` is a similar object.

File: torch/csrc/jit/test_jit.cpp [hunk 14]
```diff
@@ -246,7 +247,204 @@ void internedStringsTests () {
...
+
+    auto lstm_g = build_lstm();
+    Function  lstm_function(lstm_g);
+    std::vector<at::Tensor> outputs;
+    Interpreter lstm_interp(lstm_function);
+    lstm_interp.runOneStage({input[0], hx, cx, w_ih, w_hh}, outputs);
+    std::tie(hx, cx) = lstm(input[0], hx, cx, w_ih, w_hh);
+
+    //std::cout << almostEqual(outputs[0],hx) << "\n";
+    JIT_ASSERT(almostEqual(outputs[0],hx));
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/test_jit.cpp]: That's a good point, I wonder if they are. I just always work with non-determinism as my assumed default.

File: torch/csrc/jit/test_jit.cpp [hunk 15]
```diff
@@ -246,7 +247,204 @@ void internedStringsTests () {
 }
 
 
+
+at::Tensor t_use(at::Tensor x) {
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/test_jit.cpp]: Yeah, I have another patch that just adds the benchmarking code right here as well. :)

File: torch/csrc/jit/interpreter.cpp (hunk 12 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter.cpp]: It is possible (though unlikely), that a final stage is just an identity function of its inputs so there won't be instructions. So this makes sure we have inserted Stage objects for each stage.

File: torch/csrc/jit/interpreter_autograd_function.h (hunk 13 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter_autograd_function.h]: interp is just a handle to a `std::shared_ptr<InterpreterImpl>` so copy constructor does not make a full copy. I'll think about whether there is a way to make this more clear.

Comment (apaszke) [reply torch/csrc/jit/interpreter_autograd_function.h]: Ah right

File: torch/csrc/jit/interpreter.cpp (hunk 12 above)
Comment (apaszke) [reply torch/csrc/jit/interpreter.cpp]: Can you please add a comment? It's not immediately obvious

File: torch/csrc/jit/interpreter.h (hunk 4 above)
Comment (zdevito) [reply torch/csrc/jit/interpreter.h]: I did Function -> Code, and Interpreter -> InterpreterState.

**Comment by zdevito** [comment]
I made the RegList stuff more type safe so you no longer have to guess Bool/Int and you have to try harder to screw up appending to a list.  I tried to do more aggressive changes like create a ContiguousListPool objects, there are some optimizations that remove indirections when accessing these lists that make these changes difficult to encapsulate.
